File: exercises/day11-AWorldWithoutMock/scala/src/main/scala/service/ToyProductionService.scala
package service

import domain.{Toy, ToyRepository}
import domain.ToyState._

class ToyProductionService(repository: ToyRepository, notificationService: NotificationService) {

  def assignToyToElf(toyName: String): Unit = {
    val toy = repository.findByName(toyName)
    if (toy != null && toy.state == UNASSIGNED) {
      toy.state = IN_PRODUCTION
      repository.save(toy)
      notificationService.notifyToyAssigned(toy)
    }
  }
}

File: exercises/day11-AWorldWithoutMock/typescript/src/service/ToyProductionService.ts
import { Toy, ToyState } from "../domain/Toy";
import { ToyRepository } from "../domain/ToyRepository";
import { NotificationService } from "./NotificationService";

export class ToyProductionService {
  constructor(
    private readonly repository: ToyRepository,
    private readonly notificationService: NotificationService
  ) {}

  assignToyToElf(toyName: string): void {
    const toy = this.repository.findByName(toyName);
    if (toy && toy.state === ToyState.UNASSIGNED) {
      toy.state = ToyState.IN_PRODUCTION;
      this.repository.save(toy);
      this.notificationService.notifyToyAssigned(toy);
    }
  }
}
